src/plugins/kdb/ldap/libkdb_ldap/ldap_principal2.c

```diff
@@ -1375,64 +1482,207 @@ krb5_read_tkt_policy(krb5_context context, krb5_ldap_context *ldap_context,
     return st;
 }
 
-krb5_error_code
-krb5_decode_krbsecretkey(krb5_context context, krb5_db_entry *entries,
-                         struct berval **bvalues,
-                         krb5_tl_data *userinfo_tl_data, krb5_kvno *mkvno)
+struct key_entry {
+    krb5_int16 n_key_data;
+    krb5_key_data *key_data;
+};
+
+/*
+ * Decode keys from ldap search results.
+ *
+ * Arguments:
+ *  - bvalues
+ *      The ldap search results containing the key data.
+ *  - mkvno
+ *      The master kvno that the keys were encrypted with.
+ *  - keys_out
+ *      The decoded keys in a key_entry struct. Must be freed.
+ *  - n_keys_out
+ *      The number of key_entry's in keys_out.
+ *  - total_keys_out
+ *      An optional argument that if given will be set to the total number of
+ *      keys found throughout all the entries: sum(keys_out.n_key_data)
+ *      May be NULL.
```

Comment:
I rather like it as an optional argument, and think it is simpler. No need to have some useless variable in the caller, that could just be confusing seeing it being set but never used.